atom/renderer/api/atom_api_spell_check_client.cc

```diff
@@ -155,62 +148,93 @@ void SpellCheckClient::SpellCheckText(
   base::string16 word;
   int word_start;
   int word_length;
+  std::vector<base::string16> words;
+  auto& word_map = pending_request_param_->wordmap();
   for (auto status =
            text_iterator_.GetNextWord(&word, &word_start, &word_length);
        status != SpellcheckWordIterator::IS_END_OF_TEXT;
        status = text_iterator_.GetNextWord(&word, &word_start, &word_length)) {
     if (status == SpellcheckWordIterator::IS_SKIPPABLE)
       continue;
 
-    // Found a word (or a contraction) that the spellchecker can check the
-    // spelling of.
-    if (SpellCheckWord(scope, word))
-      continue;
-
     // If the given word is a concatenated word of two or more valid words
     // (e.g. "hello:hello"), we should treat it as a valid word.
-    if (IsValidContraction(scope, word))
-      continue;
+    std::vector<base::string16> contraction_words;
+    if (!IsContraction(scope, word, &contraction_words)) {
+      blink::WebTextCheckingResult result;
+      result.location = word_start;
+      result.length = word_length;
+      words.push_back(word);
+      word_map[word].push_back(result);
+    } else {
+      // For a contraction, we want check the spellings of each individual
+      // part, but mark the entire word incorrect if any part is misspelt
+      // Hence, we use the same word_start and word_length values for every
+      // part of the contraction.
+      for (const auto& w : contraction_words) {
+        blink::WebTextCheckingResult result;
+        result.location = word_start;
+        result.length = word_length;
+        words.push_back(w);
+        word_map[w].push_back(result);
+      }
+    }
+  }
 
-    blink::WebTextCheckingResult result;
-    result.location = word_start;
-    result.length = word_length;
-    results->push_back(result);
+  // Send out all the words data to the spellchecker to check
+  SpellCheckWords(scope, words);
+}
+
+void SpellCheckClient::OnSpellCheckDone(
+    const std::vector<base::string16>& misspelt_words) {
+  std::vector<blink::WebTextCheckingResult> results;
+  auto* const completion_handler = pending_request_param_->completion();
 
-    if (stop_at_first_result)
-      return;
+  auto& word_map = pending_request_param_->wordmap();
+
+  for (const auto& word : misspelt_words) {
+    auto iter = word_map.find(word);
+    if (iter != word_map.end()) {
+      auto& words = iter->second;
```

Comment:
_(minor)_ 'words' is a confusing variable name here because they're ranges / results. I know 'results' is already taken but could something else be used here?

Reply:
Added a comment.